File: java/EmployeeManagementSystem.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class EmployeeManagementSystem {
    private List<Employee> employeeList = new ArrayList<>();

    // Object for synchronization
    private final Object lock = new Object();

    // Save data in a separate thread
    public void saveDataToFile(String fileName) {
        Runnable saveTask = () -> {
            synchronized (lock) {
                try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(fileName))) {
                    oos.writeObject(employeeList);
                    System.out.println("Data saved successfully in thread: " + Thread.currentThread().getName());
                } catch (IOException e) {
                    System.out.println("Error saving file: " + e.getMessage());
                }
            }
        };
        new Thread(saveTask).start();
    }

    // Load data in a separate thread
    public void loadDataFromFile(String fileName) {
        @SuppressWarnings("unchecked")
        Runnable loadTask = () -> {
            synchronized (lock) {
                try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(fileName))) {
                    employeeList = (List<Employee>) ois.readObject();
                    System.out.println("Data loaded successfully in thread: " + Thread.currentThread().getName());
                } catch (IOException | ClassNotFoundException e) {
                    System.out.println("Error loading file: " + e.getMessage());
                }
            }
        };
        new Thread(loadTask).start();
    }

    // Load data from a provided list
    public void loadDataFromList(List<Employee> list) {
        synchronized (lock) {
            this.employeeList = new ArrayList<>(list);
        }
    }

    // Add employee
    public void addEmployee(Employee e) {
        synchronized (lock) {
            employeeList.add(e);
            System.out.println("Added employee: " + e.getFullName());
        }
    }

    // Remove employee
    public void removeEmployee(Employee e) {
        synchronized (lock) {
            if (employeeList.remove(e)) {
                System.out.println("Removed employee: " + e.getFullName());
            } else {
                System.out.println("Employee not found: " + e.getFullName());
            }
        }
    }

    // Get a snapshot of employee list
    public List<Employee> getEmployeeList() {
        synchronized (lock) {
            return new ArrayList<>(employeeList);
        }
    }
}

File: csharp/EmployeeManagementSystem.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading;

public class EmployeeManagementSystem
{
    private List<Employee> employeeList = new();
    private readonly object lockObj = new();

    public void SaveDataToFile(string fileName)
    {
        Thread saveThread = new(() =>
        {
            lock (lockObj)
            {
                try
                {
                    var options = new JsonSerializerOptions
                    {
                        WriteIndented = true,
                        IncludeFields = true,
                        ReferenceHandler = System.Text.Json.Serialization.ReferenceHandler.Preserve // handles circular references
                    };
                    string json = JsonSerializer.Serialize(employeeList, options);
                    File.WriteAllText(fileName, json);
                    Console.WriteLine($"Data saved successfully in thread: {Thread.CurrentThread.ManagedThreadId}");
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error saving file: {e.Message}");
                }
            }
        });
        saveThread.Start();
    }

    public void LoadDataFromFile(string fileName)
    {
        Thread loadThread = new(() =>
        {
            lock (lockObj)
            {
                try
                {
                    string json = File.ReadAllText(fileName);
                    var options = new JsonSerializerOptions
                    {
                        IncludeFields = true,
                        ReferenceHandler = System.Text.Json.Serialization.ReferenceHandler.Preserve
                    };
                    employeeList = JsonSerializer.Deserialize<List<Employee>>(json, options) ?? new List<Employee>();
                    Console.WriteLine($"Data loaded successfully in thread: {Thread.CurrentThread.ManagedThreadId}");
                }
                catch (Exception e)
                {
                    Console.WriteLine($"Error loading file: {e.Message}");
                }
            }
        });
        loadThread.Start();
    }

    public void LoadDataFromList(List<Employee> list)
    {
        lock (lockObj)
        {
            employeeList = new List<Employee>(list);
        }
    }

    public void AddEmployee(Employee e)
    {
        lock (lockObj)
        {
            employeeList.Add(e);
            Console.WriteLine($"Added employee: {e.GetFullName()}");
        }
    }

    public void RemoveEmployee(Employee e)
    {
        lock (lockObj)
        {
            if (employeeList.Remove(e))
                Console.WriteLine($"Removed employee: {e.GetFullName()}");
            else
                Console.WriteLine($"Employee not found: {e.GetFullName()}");
        }
    }

    public List<Employee> GetEmployeeList()
    {
        lock (lockObj)
        {
            return new List<Employee>(employeeList);
        }
    }
}
